eth/api_tracer.go

```diff
@@ -390,6 +390,13 @@ func (api *PrivateDebugAPI) TraceBlockFromFile(ctx context.Context, file string,
 	}
 	return api.TraceBlock(ctx, blob, config)
 }
+func (api *PrivateDebugAPI) TraceBadBlock(ctx context.Context, index int, config *TraceConfig) ([]*txTraceResult, error) {
+	if len := len(api.eth.blockchain.BadBlocks()); index < len {
+		block := api.eth.blockchain.BadBlocks()[index]
+		return api.traceBlock(ctx, block, config)
+	}
```

Comment:
This is racey. You check the length first, then later index using it, but retrieve the list again, which might have changed in between. (also please don't use `len` as a variable, it's a keyword).

```go
	if blocks := api.eth.blockchain.BadBlocks(); index < len(blocks) {
		return api.traceBlock(ctx, blocks[index], config)
	}
```